File: src/main/java/net/minecraft/client/particle/Particle.java
package net.minecraft.client.particle;

import com.mojang.nbt.CompoundTag;
import net.minecraft.client.renderer.Tesselator;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.Level;
import util.Mth;

public class Particle extends Entity {
   protected int tex;
   protected float uo;
   protected float vo;
   protected int age = 0;
   protected int lifetime = 0;
   protected float size;
   protected float gravity;
   protected float rCol;
   protected float gCol;
   protected float bCol;
   public static double xOff;
   public static double yOff;
   public static double zOff;

   public Particle(Level level, double x, double y, double z, double xa, double ya, double za) {
      super(level);
      this.setSize(0.2F, 0.2F);
      this.heightOffset = this.bbHeight / 2.0F;
      this.setPos(x, y, z);
      this.rCol = this.gCol = this.bCol = 1.0F;
      this.xd = xa + (double)((float)(Math.random() * (double)2.0F - (double)1.0F) * 0.4F);
      this.yd = ya + (double)((float)(Math.random() * (double)2.0F - (double)1.0F) * 0.4F);
      this.zd = za + (double)((float)(Math.random() * (double)2.0F - (double)1.0F) * 0.4F);
      float speed = (float)(Math.random() + Math.random() + (double)1.0F) * 0.15F;
      float dd = Mth.sqrt(this.xd * this.xd + this.yd * this.yd + this.zd * this.zd);
      this.xd = this.xd / (double)dd * (double)speed * (double)0.4F;
      this.yd = this.yd / (double)dd * (double)speed * (double)0.4F + (double)0.1F;
      this.zd = this.zd / (double)dd * (double)speed * (double)0.4F;
      this.uo = this.random.nextFloat() * 3.0F;
      this.vo = this.random.nextFloat() * 3.0F;
      this.size = (this.random.nextFloat() * 0.5F + 0.5F) * 2.0F;
      this.lifetime = (int)(4.0F / (this.random.nextFloat() * 0.9F + 0.1F));
      this.age = 0;
      this.makeStepSound = false;
   }

   public Particle setPower(float power) {
      this.xd *= (double)power;
      this.yd = (this.yd - (double)0.1F) * (double)power + (double)0.1F;
      this.zd *= (double)power;
      return this;
   }

   public Particle scale(float scale) {
      this.setSize(0.2F * scale, 0.2F * scale);
      this.size *= scale;
      return this;
   }

   protected void defineSynchedData() {
   }

   public void tick() {
      this.xo = this.x;
      this.yo = this.y;
      this.zo = this.z;
      if (this.age++ >= this.lifetime) {
         this.remove();
      }

      this.yd -= 0.04 * (double)this.gravity;
      this.move(this.xd, this.yd, this.zd);
      this.xd *= (double)0.98F;
      this.yd *= (double)0.98F;
      this.zd *= (double)0.98F;
      if (this.onGround) {
         this.xd *= (double)0.7F;
         this.zd *= (double)0.7F;
      }

   }

   public void render(Tesselator t, float a, float xa, float ya, float za, float xa2, float za2) {
      float u0 = (float)(this.tex % 16) / 16.0F;
      float u1 = u0 + 0.0624375F;
      float v0 = (float)(this.tex / 16) / 16.0F;
      float v1 = v0 + 0.0624375F;
      float r = 0.1F * this.size;
      float x = (float)(this.xo + (this.x - this.xo) * (double)a - xOff);
      float y = (float)(this.yo + (this.y - this.yo) * (double)a - yOff);
      float z = (float)(this.zo + (this.z - this.zo) * (double)a - zOff);
      float br = this.getBrightness(a);
      t.color(this.rCol * br, this.gCol * br, this.bCol * br);
      t.vertexUV((double)(x - xa * r - xa2 * r), (double)(y - ya * r), (double)(z - za * r - za2 * r), (double)u1, (double)v1);
      t.vertexUV((double)(x - xa * r + xa2 * r), (double)(y + ya * r), (double)(z - za * r + za2 * r), (double)u1, (double)v0);
      t.vertexUV((double)(x + xa * r + xa2 * r), (double)(y + ya * r), (double)(z + za * r + za2 * r), (double)u0, (double)v0);
      t.vertexUV((double)(x + xa * r - xa2 * r), (double)(y - ya * r), (double)(z + za * r - za2 * r), (double)u0, (double)v1);
   }

   public int getParticleTexture() {
      return 0;
   }

   public void addAdditonalSaveData(CompoundTag entityTag) {
   }

   public void readAdditionalSaveData(CompoundTag tag) {
   }
}

File: mc-cpp/src/particle/Particle.cpp
#include "particle/Particle.hpp"
#include "world/Level.hpp"
#include "util/Mth.hpp"
#include <cmath>

namespace mc {

// Static members
double Particle::xOff = 0.0;
double Particle::yOff = 0.0;
double Particle::zOff = 0.0;

Particle::Particle(Level* level, double x, double y, double z, double xa, double ya, double za)
    : level(level)
    , random(std::random_device{}())
{
    bbWidth = 0.2f;
    bbHeight = 0.2f;
    setPos(x, y, z);

    // Initialize color to white
    rCol = gCol = bCol = 1.0f;

    // Add randomness to velocity (matching Java)
    auto randFloat = [this]() { return static_cast<float>(random()) / random.max(); };

    xd = xa + (randFloat() * 2.0f - 1.0f) * 0.4f;
    yd = ya + (randFloat() * 2.0f - 1.0f) * 0.4f;
    zd = za + (randFloat() * 2.0f - 1.0f) * 0.4f;

    // Normalize and scale velocity
    float speed = (randFloat() + randFloat() + 1.0f) * 0.15f;
    float dd = Mth::sqrt(static_cast<float>(xd * xd + yd * yd + zd * zd));
    xd = xd / dd * speed * 0.4;
    yd = yd / dd * speed * 0.4 + 0.1;
    zd = zd / dd * speed * 0.4;

    // Random texture offset
    uo = randFloat() * 3.0f;
    vo = randFloat() * 3.0f;

    // Random size
    size = (randFloat() * 0.5f + 0.5f) * 2.0f;

    // Random lifetime (4-40 ticks)
    lifetime = static_cast<int>(4.0f / (randFloat() * 0.9f + 0.1f));
    age = 0;
}

Particle& Particle::setPower(float power) {
    xd *= power;
    yd = (yd - 0.1) * power + 0.1;
    zd *= power;
    return *this;
}

Particle& Particle::scale(float s) {
    bbWidth = 0.2f * s;
    bbHeight = 0.2f * s;
    size *= s;
    return *this;
}

void Particle::tick() {
    prevX = x;
    prevY = y;
    prevZ = z;

    if (age++ >= lifetime) {
        remove();
        return;
    }

    // Apply gravity
    yd -= 0.04 * gravity;

    // Move
    move(xd, yd, zd);

    // Apply air resistance
    xd *= 0.98;
    yd *= 0.98;
    zd *= 0.98;

    // Apply ground friction
    if (onGround) {
        xd *= 0.7;
        zd *= 0.7;
    }
}

void Particle::move(double dx, double dy, double dz) {
    if (!level) {
        x += dx;
        y += dy;
        z += dz;
        updateBoundingBox();
        return;
    }

    double originalDx = dx;
    double originalDy = dy;
    double originalDz = dz;

    // Get collision boxes
    AABB expandedBB = bb.expand(dx, dy, dz);
    auto boxes = level->getCollisionBoxes(nullptr, expandedBB);

    // Adjust Y movement
    for (const auto& box : boxes) {
        dy = box.clipYCollide(bb, dy);
    }
    bb = bb.move(0.0, dy, 0.0);

    // Adjust X movement
    for (const auto& box : boxes) {
        dx = box.clipXCollide(bb, dx);
    }
    bb = bb.move(dx, 0.0, 0.0);

    // Adjust Z movement
    for (const auto& box : boxes) {
        dz = box.clipZCollide(bb, dz);
    }
    bb = bb.move(0.0, 0.0, dz);

    // Update position from bounding box
    x = (bb.x0 + bb.x1) / 2.0;
    y = bb.y0;
    z = (bb.z0 + bb.z1) / 2.0;

    // Check for ground collision
    onGround = originalDy != dy && originalDy < 0.0;

    // Stop velocity on collision
    if (originalDx != dx) xd = 0.0;
    if (originalDy != dy) yd = 0.0;
    if (originalDz != dz) zd = 0.0;
}

void Particle::setPos(double newX, double newY, double newZ) {
    x = newX;
    y = newY;
    z = newZ;
    prevX = newX;
    prevY = newY;
    prevZ = newZ;
    updateBoundingBox();
}

void Particle::updateBoundingBox() {
    float w = bbWidth / 2.0f;
    float h = bbHeight;
    bb = AABB(x - w, y, z - w, x + w, y + h, z + w);
}

float Particle::getBrightness() const {
    if (!level) return 1.0f;
    return level->getBrightness(
        static_cast<int>(std::floor(x)),
        static_cast<int>(std::floor(y)),
        static_cast<int>(std::floor(z))
    );
}

} // namespace mc
